tests/integration/security/jwt_test.go

```diff
@@ -62,332 +61,275 @@ func TestRequestAuthentication(t *testing.T) {
-									Headers: map[string][]string{
-										authHeaderKey: {"Bearer " + jwt.TokenExpired},
-									},
-									Count: callCount,
-								},
-								DestClusters: dest.Clusters(),
+						ExpectResponseCode: response.StatusCodeOK,
+					},
+					{
+						Name:   "invalid-jwks-valid-token-noauthz",
+						Config: "invalid-jwks",
+						CallOpts: echo.CallOptions{
+							PortName: "http",
+							Scheme:   scheme.HTTP,
+							Headers: map[string][]string{
+								authHeaderKey: {"Bearer " + jwt.TokenIssuer1},
 							},
-							ExpectResponseCode: response.StatusUnauthorized,
+							Count: callCount,
 						},
-						{
-							Name:   "invalid-jwks-no-token-noauthz",
-							Config: "invalid-jwks",
-							Request: connection.Checker{
-								From: client,
-								Options: echo.CallOptions{
-									Target:   dest[0],
-									PortName: "http",
-									Scheme:   scheme.HTTP,
-									Count:    callCount,
-								},
-								DestClusters: dest.Clusters(),
+						ExpectResponseCode: response.StatusUnauthorized,
+					},
+					{
+						Name:   "invalid-jwks-expired-token-noauthz",
+						Config: "invalid-jwks",
+						CallOpts: echo.CallOptions{
+							PortName: "http",
+							Scheme:   scheme.HTTP,
+							Headers: map[string][]string{
+								authHeaderKey: {"Bearer " + jwt.TokenExpired},
 							},
-							ExpectResponseCode: response.StatusCodeOK,
+							Count: callCount,
 						},
-					}
-					for _, c := range testCases {
-						t.NewSubTest(c.Name).Run(func(t framework.TestContext) {
+						ExpectResponseCode: response.StatusUnauthorized,
+					},
+					{
+						Name:   "invalid-jwks-no-token-noauthz",
+						Config: "invalid-jwks",
+						CallOpts: echo.CallOptions{
+							PortName: "http",
+							Scheme:   scheme.HTTP,
+							Count:    callCount,
+						},
+						ExpectResponseCode: response.StatusCodeOK,
+					},
+				}
+				for _, c := range testCases {
+					echotest.New(t, apps.All).
+						SetupForPair(func(t framework.TestContext, src, dst echo.Instances) error {
 							if c.Config != "" {
-								policy := tmpl.EvaluateOrFail(t,
-									file.AsStringOrFail(t, fmt.Sprintf("testdata/requestauthn/%s.yaml.tmpl", c.Config)), namespaceTmpl)
-								t.Config().ApplyYAMLOrFail(t, ns.Name(), policy)
-								util.WaitForConfig(t, policy, ns)
+								policy := yml.MustApplyNamespace(t, tmpl.MustEvaluate(
+									file.AsStringOrFail(t, fmt.Sprintf("testdata/requestauthn/%s.yaml.tmpl", c.Config)),
+									map[string]string{
+										"Namespace": ns.Name(),
+										"dst":       dst[0].Config().Service,
+									},
+								), ns.Name())
+								return t.Config().ApplyYAML(ns.Name(), policy)
 							}
-
-							retry.UntilSuccessOrFail(t, c.CheckAuthn, echo.DefaultCallRetryOptions()...)
+							return nil
+						}).
+						From(
+							echotest.SingleSimplePodServiceAndAllSpecial(),
+							echotest.Not(func(instances echo.Instances) echo.Instances { return instances.Match(echo.IsHeadless()) }),
```

Comment:
can we merge these to a single `echotest.Not(...)`?

Reply:
I don't have a good idea about merging these matchers into a single one.
Maybe we can write it as echotest.Not(func(...){...}).Not(func(...){...}).Not(func(...){...})...) That seems to be harder to read.

Reply:
1. Making `Not` varadic, could be should be clear what the behavior is: it could be `!(a || b || c)` (match if none of the conditions pass) or `!a || !b || !c` (match if any one of the conditions fail). The behavior of using separate `Not`'s is the former. I'm not sure I want to introduce this potentially confusing abstraction.


2. To make this less verbose, everyone should feel free to add more filters in `echotest/filters.go`

Example:

```go
func HeadlessServices(instances echo.Instances) {
  return FilterMatch(echo.IsHeadless)
}
```

then use it like 

```go
echo.Not(echotest.IsHeadless)
```

Reply:
Also, when reachability is a concern, check out `ReachableDestinations` often to see if it covers your cases.